File: src/main/java/me/testaccount666/serversystem/commands/executables/warp/TabCompleterWarp.java
package me.testaccount666.serversystem.commands.executables.warp;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.executables.warp.manager.Warp;
import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.command.Command;

import java.util.List;
import java.util.Optional;

public class TabCompleterWarp implements ServerSystemTabCompleter {

    @Override
    public Optional<List<String>> tabComplete(User commandSender, Command command, String label, String... arguments) {
        var commandName = command.getName();
        if (!commandName.equalsIgnoreCase("warp") && !commandName.equalsIgnoreCase("deletewarp")) return Optional.empty();

        var permissionPath = switch (commandName) {
            case "warp" -> "Warp.Use";
            case "deletewarp" -> "Warp.Delete";
            case "setwarp" -> "Warp.Set";
            default -> null;
        };

        if (!PermissionManager.hasCommandPermission(commandSender, permissionPath, false)) return Optional.of(List.of());

        var potentialCompletions = ServerSystem.Instance.getWarpManager().getWarps().stream().map(Warp::getName).toList();
        var completions = potentialCompletions.stream().filter(completion -> completion.toLowerCase().startsWith(arguments[0].toLowerCase())).toList();

        return Optional.of(completions);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/warp/TabCompleterWarp.kt
package me.testaccount666.serversystem.commands.executables.warp

import me.testaccount666.serversystem.ServerSystem.Companion.instance
import me.testaccount666.serversystem.commands.executables.warp.manager.WarpManager
import me.testaccount666.serversystem.commands.interfaces.ServerSystemTabCompleter
import me.testaccount666.serversystem.managers.PermissionManager.hasCommandPermission
import me.testaccount666.serversystem.userdata.User
import org.bukkit.command.Command

class TabCompleterWarp : ServerSystemTabCompleter {
    override fun tabComplete(commandSender: User, command: Command, label: String, vararg arguments: String): List<String>? {
        val commandName = command.name
        if (!commandName.equals("warp", true)
            && !commandName.equals("deletewarp", true)) return null

        val permissionPath = when (commandName) {
            "warp" -> "Warp.Use"
            "deletewarp" -> "Warp.Delete"
            "setwarp" -> "Warp.Set"
            else -> null
        }

        if (!hasCommandPermission(commandSender, permissionPath, false)) return listOf()

        val potentialCompletions = instance.registry.getService<WarpManager>().warps.map { it.name }
        return potentialCompletions.filter { it.startsWith(arguments[0], true) }
    }
}
